blockstack/blockstack-browser
# Make Onboarding More Failure-Resistant

Closes #1506. Closes #1524. Closes #1548.

### What This Does

This adds failure handling to most aspects of onboarding. You'll now both be informed of, and will be able to continue through onboarding bug-free if the following happen:

* You're unable to check the availability of your username
* Your username registration fails
* Your recovery email fails to send

Email requests are now done in parallel, and cap out at 10 seconds on slow connections.

An implication of this change is that users can now end up in the browser after onboarding without a username. While this was already a use case we supported, it'll be more important now that we behave without issue in that case (Hopefully we suss out all issues after #1547 is complete.)

In addition to the above features and changes, this PR adds the [`reapop` library](https://github.com/LouisBarranqueiro/reapop) as per @aulneau's suggestion for showing error messages. This should be useful in other parts of the browser. Here's an example of how to use:

```.jsx
import { notify } from 'reapop'

class MyComponent extends React.Component {
  componentDidMount() {
    this.props.notify({
      title: 'Hello!',
      // ...
    })
  }
}

connect(mapStateToProps, () => {
  notify
})(MyComponent)
```

### Steps to Test

The easiest way to test all of these cases is to use [Chrome's "Request Blocking" feature](https://developers.google.com/web/updates/2017/04/devtools-release-notes#block-requests).

* **Failed Username Check**
  1. Go through onboarding until the username step
  2. Block requests to `registrar.blockstack.org`
  3. Attempt to check on a username
  4. Confirm you get an error and the skip button appears
  5. Skip the step, confirm onboarding continues correctly and shows no username at the end
* **Failed Username Registration**
  1. Go through onboarding until the username step
  2. Attempt to check on a username succesfully
  3. Block requests to `registrar.blockstack.org`
  4. Attempt to register the username
  5. Confirm that after it finishes, you see a popup informing you it failed
  6. Confirm onboarding continues correctly and shows no username at the end
* **Emails failed to send**
  1. Go through onboarding until the username step
  2. Block requests to `browser-api.blockstack.org`
  3. Complete the step, confirm that the next step you see informs you of the failure to send the email
  4. Unblock requests and click "resend"
  5. Confirm that the email successfully resends
  6. Confirm onboarding continues correctly

## Screenshots

### Failed to check username availability (w/ skip button)

<img width="479" alt="screen shot 2018-08-07 at 3 01 15 pm" src="https://user-images.githubusercontent.com/649992/43796622-c3dde826-9a52-11e8-84d4-556f99045771.png">


### Failed to register username

<img width="831" alt="screen shot 2018-08-07 at 3 02 23 pm" src="https://user-images.githubusercontent.com/649992/43796679-ee3029b8-9a52-11e8-96d7-96b5f8dc0eb2.png">


### Email failed to send

<img width="483" alt="screen shot 2018-08-07 at 3 04 07 pm" src="https://user-images.githubusercontent.com/649992/43796763-29155724-9a53-11e8-8b87-2195e2e24bc4.png">



### End screen after failed / skipped username registration

<img width="479" alt="screen shot 2018-08-07 at 2 49 15 pm" src="https://user-images.githubusercontent.com/649992/43796479-61eee0f2-9a52-11e8-8862-9fcfefdd984b.png">
